rclpy/src/rclpy/_rclpy_action.c

```diff
@@ -0,0 +1,885 @@
+ * \param[in] pyresult_service_qos QoSProfile Python object for the result service.
+ * \param[in] pycancel_service_qos QoSProfile Python object for the cancel service.
+ * \param[in] pyfeedback_qos QoSProfile Python object for the feedback subscriber.
+ * \param[in] pystatus_qos QoSProfile Python object for the status subscriber.
+ * \return capsule and memory address, or
+ * \return NULL on failure.
+ */
+static PyObject *
+rclpy_action_create_client(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pynode;
+  PyObject * pyaction_type;
+  PyObject * pyaction_name;
+  PyObject * pygoal_service_qos;
+  PyObject * pyresult_service_qos;
+  PyObject * pycancel_service_qos;
+  PyObject * pyfeedback_topic_qos;
+  PyObject * pystatus_topic_qos;
+
+  int parse_tuple_result = PyArg_ParseTuple(
+    args,
+    "OOOOOOOO",
+    &pynode,
+    &pyaction_type,
+    &pyaction_name,
+    &pygoal_service_qos,
+    &pyresult_service_qos,
+    &pycancel_service_qos,
+    &pyfeedback_topic_qos,
+    &pystatus_topic_qos);
+
+  if (!parse_tuple_result) {
+    return NULL;
+  }
+
+  if (!PyUnicode_Check(pyaction_name)) {
+    PyErr_Format(PyExc_TypeError, "Argument pyaction_name is not a PyUnicode object");
+    return NULL;
+  }
+
+  char * action_name = (char *)PyUnicode_1BYTE_DATA(pyaction_name);
+
+  rcl_node_t * node = (rcl_node_t *)PyCapsule_GetPointer(pynode, "rcl_node_t");
+
+  PyObject * pymetaclass = PyObject_GetAttrString(pyaction_type, "__class__");
+
+  PyObject * pyts = PyObject_GetAttrString(pymetaclass, "_TYPE_SUPPORT");
+
+  rosidl_action_type_support_t * ts =
+    (rosidl_action_type_support_t *)PyCapsule_GetPointer(pyts, NULL);
+
+  rcl_action_client_options_t action_client_ops = rcl_action_client_get_default_options();
+
+  OPTIONS_COPY_QOS_PROFILE(action_client_ops, goal_service_qos);
+  OPTIONS_COPY_QOS_PROFILE(action_client_ops, result_service_qos);
+  OPTIONS_COPY_QOS_PROFILE(action_client_ops, cancel_service_qos);
+  OPTIONS_COPY_QOS_PROFILE(action_client_ops, feedback_topic_qos);
+  OPTIONS_COPY_QOS_PROFILE(action_client_ops, status_topic_qos);
+
+  rcl_action_client_t * action_client =
+    (rcl_action_client_t *)PyMem_Malloc(sizeof(rcl_action_client_t));
+  *action_client = rcl_action_get_zero_initialized_client();
+  rcl_ret_t ret = rcl_action_client_init(
+    action_client,
+    node,
+    ts,
+    action_name,
+    &action_client_ops);
+  if (ret != RCL_RET_OK) {
+    if (ret == RCL_RET_ACTION_NAME_INVALID) {
+      PyErr_Format(PyExc_ValueError,
+        "Failed to create action client due to invalid topic name '%s': %s",
+        action_name, rcl_get_error_string().str);
+    } else {
+      PyErr_Format(PyExc_RuntimeError,
+        "Failed to create action client: %s", rcl_get_error_string().str);
+    }
+    PyMem_Free(action_client);
+    rcl_reset_error();
+    return NULL;
+  }
+
+  return PyCapsule_New(action_client, "rcl_action_client_t", NULL);
```

Comment:
I've added a `__del__` method to `ActionClient` to ensure the object is destroyed during garbage collection.

I can add a PyCapsule destructor. But I guess we should choose one strategy for destruction to avoid a double-free?

Reply:
Another issue that came up is that we should ensure that the `Node` is destroyed *after* the action client, since the rcl finalize functions require a reference to the node. What do you think about adding a `destroy()` method to the `Waitable` interface for `Node` to call or a `Destroyable` interface? I'm not sure if we have to worry about ABI/API issues with the former.

Reply:
I've added a check in the destroy method to avoid a potential segfault if the action client is destroyed after the node: f2b238c

Reply:
> Another issue that came up is that we should ensure that the Node is destroyed after the action client

There are a bunch of memory issues with nodes and node entities in `rclpy`. What I'd like to see is too big for this PR (if something needs a node, it should hold a reference to the node rather than the other way around), so I'm fine with setting the capsule pointer to a sentinel value `PyNone` for now.